onnxruntime/core/providers/coreml/builders/impl/argmax_op_builder.cc

```diff
@@ -0,0 +1,115 @@
+#include "core/providers/coreml/builders/model_builder.h"
+#include "core/providers/coreml/builders/op_builder_factory.h"
+
+#include "base_op_builder.h"
+
+namespace onnxruntime {
+namespace coreml {
+
+class ArgMaxOpBuilder : public BaseOpBuilder {
+  // Add operator related
+ private:
+  Status AddToModelBuilderImpl(ModelBuilder& model_builder, const Node& node,
+                               const logging::Logger& logger) const override ORT_MUST_USE_RESULT;
+
+  // Operator support related
+ private:
+  bool IsOpSupportedImpl(const Node& node, OpBuilderInputParams& input_params,
+                         const logging::Logger& logger) const override;
+};
+
+// Add operator related
+
+Status ArgMaxOpBuilder::AddToModelBuilderImpl(ModelBuilder& model_builder,
+                                              const Node& node,
+                                              const logging::Logger& /* logger */) const {
+  std::unique_ptr<COREML_SPEC::NeuralNetworkLayer> layer = CreateNNLayer(node);
+  const auto& graph_viewer = model_builder.GetGraphViewer();
+
+  NodeAttrHelper helper(node);
+  const auto axis = helper.Get("axis", 0);
+  const auto keep_dims = helper.Get("keep_dims", 1);
+  bool removedim = false;
+  if (keep_dims != 1) {
+    removedim = true;
+  }
+
+  auto* coreml_argmax = layer->mutable_argmax();
+  coreml_argmax->set_axis(axis);
+  coreml_argmax->set_removedim(removedim);
+
+  // Get ArgMax's next node(Cast)'s outputdefs
+  auto it = node.OutputEdgesBegin();
+  const auto* succ_node(graph_viewer.GetNode(it->GetNode().Index()));
+
+  // Skip the cast's input/argmax's output
+  *layer->mutable_input()->Add() = node.InputDefs()[0]->Name();
+  *layer->mutable_output()->Add() = succ_node->OutputDefs()[0]->Name();
+
+  model_builder.AddLayer(std::move(layer));
+  return Status::OK();
+}
+
+// Operator support related
+
+bool ArgMaxOpBuilder::IsOpSupportedImpl(const Node& node, OpBuilderInputParams& input_params,
+                                        const logging::Logger& logger) const {
+  // Check if Argmax's output is the graph output
+  const auto& graph_output_list = input_params.graph_viewer.GetOutputs();
+  std::unordered_set<const NodeArg*> graph_outputs(graph_output_list.cbegin(), graph_output_list.cend());
+  const auto& output_defs = node.OutputDefs();
+  for (const auto* output_def : output_defs) {
+    if (graph_outputs.count(output_def) != 0) {
+      LOGS(logger, VERBOSE) << "Case: ArgMax's output is the graph output: Not supported.";
+      return false;
+    }
+  }
+
+  std::vector<size_t> succ_node_indices;
+  for (auto it = node.OutputEdgesBegin(), end = node.OutputEdgesEnd(); it != end; ++it) {
+    succ_node_indices.push_back(it->GetNode().Index());
+  }
+
+  // Case where argmax has multiple succeeding nodes is not supported
+  if (succ_node_indices.size() > 1) {
+    LOGS(logger, VERBOSE) << "Case - [ArgMax] has multiple succedding nodes: Not supported.";
+    return false;
```

Comment:
nit: confusing error message for user as they're not aware of the 'preceeding/succeeding' context so this may sound like there are multiple successful nodes. 'multiple nodes consuming its output' may be clearer.

Is this just for simplicity for now? If all down stream nodes could be executed on coreml this doesn't matter does it?

Reply:
> Is this just for simplicity for now? If all down stream nodes could be executed on coreml this doesn't matter does it?

For this, I thought if there's a case where ArgMax has two succeeding nodes for example, Cast & Another op[not supported].Then in the current coreml partition logic, if it hits that node not supported, then the partition will end and thus cause the cast node not properly being added into the partition that has argmax. 

And if it's the case where ArgMax has two succeeding nodes, cast and another node [supported] , then the partition can be correct. However, since the argmax support currently only works with the int64 output fused by the cast node, the input/output int64 type between argmax and the other supported op can't be handled. (That's my understanding not sure if there's better way to handle multiple succeeding nodes.)